### DISCUSSION TITLE
Training LLaMA 3 on TPUs | How To Scale Your Model

### DISCUSSION THREAD

**Discussion by jacobaustin123** [jax-ml/scaling-book#8] [General]
Training LLAMA on TPUs!

**Comment by Serega6678** [comment]
Thank you for this section!
There is a type in the "Question: LLaMA 3-70B was pretrained with a batch size of about 4M tokens..."
In the table in the answer, the Total is incorrect, after the table Total seems correct

Thanks for this wonderful book again!

**Reply by jacobaustin123** [reply]
Fixed!

**Comment by dhaya** [comment]
In the answer to the 4th question:
> 44 days * 8960 / 171

The denominator should be 225,

**Reply by dhaya** [reply]
Also the same question mentions "gradients 4 times per layer", but the answer assumes 3.

**Reply by jacobaustin123** [reply]
Fixed. Thank you!

**Comment by tengyifei** [comment]
What's the meaning of "number of gradient checkpoints" -- is it the factor of the total size of residuals compared to the size of the layer input?

**Reply by jacobaustin123** [reply]
https://jax-ml.github.io/scaling-book/transformers/#gradient-checkpointing should explain it.

**Comment by kerrywang** [comment]
When computing the T_math, should we, in practice, also take into consideration of TPU utilization? 

eg: in the LLaMA 70B, we assumed 40% FLOPs utilization, should we instead compute $$T_{math} =  \frac {B \cdot D \cdot F} {N \cdot C \cdot \text{FLOPS UTILIZATION}}$$

**Reply by jacobaustin123** [reply]
Generally no, while LLAMA 70B might have an aggregate FLOPs utilization of 40%, the matmuls themselves will have very good FLOPs utilization (upwards of 90% often). The low overall utilization (esp during decoding) comes from things like attention, layer-norms, and imperfect ICI overlapping.

In other words, you should include that discount factor when saying `total llama training time = total llama FLOPs / (total llama FLOPs/s * 0.4)` but not when talking about a specific matmul.

**Comment by henryhmko** [comment]
Thanks for making an amazing resource! I just noticed two small typos:

1. Question 4
> *1752 days to train. That's 6 and a half years.*

To "4.8 years"

2. How to shard Llama3-70B intro
> *with 4M token batch size (1024 sequences of length 8192 per batch)*

To "1024 sequences of length 4096 per batch" from the Llama3 paper

**Reply by jacobaustin123** [reply]
Fixed. Thanks for flagging these.

**Comment by zhiyisun** [comment]
In the last question of How to shard LLaMA 3-70B for training, the answer mentions "4-way **model** parallelism", but the takeaway specifies 4-way tensor parallelism. Should this be updated to 'tensor parallelism' for consistency, given the sharding along F?

**Reply by jacobaustin123** [reply]
Yes. The terms are a bit ambiguous, but I tend to think of tensor parallelism as specifically this method while model sharding includes any method that shards the weights (even e.g. FSDP). Updated.

**Comment by lhy-loveworld** [comment]
Hi! Regarding `train LLaMA 3-70B with only FSDP on 8960 chips`, is that feasible, given that FSDP shards the d_model dimension and LLaMA 3-70B's d_model is merely 8192 which is less than the number of chips?

**Comment by lhy-loveworld** [comment]
And the huggingface page for the LLaMA 3-405B model has been deleted

**Comment by BrownianNotion** [comment]
The optimal combination for llama 70B uses 1024 FSDP + 2 way CP + 4 way TP. This gives 8192 chips, and results in 8960-8192=768 leftover chips. How would we deal with this in reality? Would we:
1. Accept that 768 chips are idle and use them for other jobs. In this case, we'd need to also update our training time estimates for a lower number of chips.
2. Modify the parallelism strategy to be a be a non-power of 2, eg. 8960/8 = 1120 way FSDP (unfortunately this also exceeds the number of sequences in the batch 1024).

**Reply by jacobaustin123** [reply]
Depends what you want but probably use them for something else. I'm sure if you used them in practice you'd see _some_ win but it might not be worth it.

**Comment by xuuuge** [comment]
Hugging face link expired.

**Reply by xuuuge** [reply]
LLaMA 3-405B config link

**Reply by jacobaustin123** [reply]
I think you just need to be logged in.

**Reply by danhussey** [reply]
When I accessed the link while logged to HF, I got a 403 error, and needed to request access to the model on the model page - just so you know
Thanks for the awesome resource
